File: tests/fixtures/python/bad_handler.py
"""Deliberately broken durable handler — every rule should fire at least once.

Modeled on the failure cases AWS documents plus the shape of a real
human-in-the-loop pipeline handler.
"""

import random
import time
import uuid
from datetime import datetime

import boto3
import requests
from aws_durable_execution_sdk_python import DurableContext, durable_execution

table = boto3.resource("dynamodb").Table("orders")
AUDIT = []


@durable_execution
def lambda_handler(event, context: DurableContext):
    run_id = event["run_id"]

    # RG001 — clock outside a step
    started_at = time.time()

    # RG001 — identity outside a step; downstream steps get a different id on replay
    correlation_id = uuid.uuid4()

    # RG001 — clock via a from-import alias, which a regex checker would miss
    stamp = datetime.now()

    # RG001 — random outside a step
    jitter = random.random()

    # RG002 — external I/O outside a step, re-executed on every replay
    profile = requests.get(f"https://example.internal/profile/{run_id}")

    # RG002 — AWS SDK call outside a step
    table.put_item(Item={"id": run_id, "started": started_at})

    # RG004 — branch on a nondeterministic value; replay may take the other path
    if datetime.now().hour < 12:
        shift = "morning"
    else:
        shift = "afternoon"

    receipts = []

    # RG003 — step body mutating a captured list; silently lost on replay
    def save(_):
        receipt = table.put_item(Item={"id": run_id})
        receipts.append(receipt)
        AUDIT.append(run_id)
        return "saved"

    context.step(save, name="save-order")

    # RG003 — same bug via a lambda and a subscript write
    state = {}

    context.step(
        lambda _: state.__setitem__("done", True),
        name="mark-done",
    )

    # RG005 — dynamically constructed step name cannot be matched on resume
    context.step(lambda _: "work", name=f"process-{time.time()}")

    # RG900 — step body passed by reference, so its region is unresolved
    context.step(_helper, name="helper")

    return {"run_id": run_id, "shift": shift, "jitter": jitter, "profile": profile}


def _helper(_):
    return "helper"

File: tests/fixtures/rust/bad_handler.rs
//! Deliberately broken durable handler, in the durable-rust closure style.
//!
//! Note what is *absent*: capturing `&mut` state in a step closure. The SDK's
//! `Send + 'static` bound means that does not compile, so the whole shape that
//! JavaScript makes easy is rejected by the borrow checker before this tool
//! ever sees it. What remains is interior mutability through a shared handle.

use chrono::Utc;
use durable_lambda_core::context::DurableContext;
use std::sync::{Arc, Mutex};
use std::time::SystemTime;
use uuid::Uuid;

static mut AUDIT_COUNT: u64 = 0;

#[durable_execution]
async fn handler(event: Value, mut ctx: DurableContext) -> Result<Value, Error> {
    // RG001 -- clock outside a step
    let started = SystemTime::now();

    // RG001 -- clock via chrono, written with the crate prefix
    let stamp = chrono::Utc::now();

    // RG001 -- identity outside a step
    let correlation = Uuid::new_v4();

    // RG002 -- filesystem outside a step
    let config = std::fs::read_to_string("/tmp/config.json")?;

    // RG004 -- branch on wall-clock time
    let shift = if Utc::now().hour() < 12 { "morning" } else { "afternoon" };

    let receipts = Arc::new(Mutex::new(Vec::new()));
    let receipts_for_step = Arc::clone(&receipts);

    // RG003 -- interior mutability through a shared handle, lost on replay
    ctx.step("save_order", move || {
        let receipts = Arc::clone(&receipts_for_step);
        async move {
            receipts.lock().unwrap().push("receipt-1".to_string());
            Ok::<_, String>("saved")
        }
    })
    .await?;

    // RG005 -- operation name built from the clock
    ctx.step(&format!("process-{}", Utc::now()), || async { Ok::<_, String>(1) })
        .await?;

    // Read back outside the step: this is what makes the write above a lost
    // update rather than a harmless write-only log.
    let count = receipts.lock().unwrap().len();

    Ok(json!({ "shift": shift, "count": count, "started": started,
               "stamp": stamp, "correlation": correlation, "config": config }))
}
